### PR TITLE
[Transformations] Apply SDPA scale after MatMul(Q, K^T) per specification

### PR DESCRIPTION
**Pull request by evkotov**
## Summary

Apply SDPA scale factor unconditionally after `MatMul(Q, K^T)` per the [SDPA specification](https://docs.openvino.ai/2025/documentation/openvino-ir-format/operation-sets/operation-specs/sequence/scaled-dot-product-attention.html): `attn_weight = Q @ K^T * scale`.

This replaces the previous conditional logic (`can_move_scale_after_matmul`) that applied scale either after MatMul or before MatMul on Q depending on shape analysis.

## Details

When `ov.convert_model()` processes transformer models, `SDPAFusion` creates SDPA nodes during MOC transformations. The CPU plugin then decomposes these back to `MatMul + Softmax + MatMul` via `ScaledDotProductAttentionDecomposition`. The previous decomposition sometimes applied scale to Q before MatMul (`(Q * scale) @ K^T`), which changed the FP32 computation order compared to the original graph. While mathematically equivalent, the rounding differences accumulate through residual connections across transformer layers (up to 0.91 max_diff on RFDetr with 14 attention blocks).

The fix simplifies the decomposition to always compute `(Q @ K^T) * scale`, which:
- Matches the SDPA specification exactly
- Post-MatMul scalar Multiply fuses into oneDNN MatMul primitive (no performance regression)
- Produces consistent behavior regardless of shape dynamism

### Tickets:
 - 181409
 - 180477

### CODE REVIEW

File: src/common/transformations/src/transformations/common_optimizations/common_optimizations.cpp [hunk 1]
```diff
@@ -138,6 +140,16 @@ bool ov::pass::CommonOptimizations::run_on_model(const std::shared_ptr<ov::Model
     // before CommonOptimization pipeline execution
     REGISTER_PASS(manager, MOCTransformations, true, false)
 
+    // SDPAFusion and SDPAScaleFusion are registered here (in CommonOptimizations) rather than
+    // inside MOCTransformations so they only run during compile_model(), not during convert_model().
+    // This allows each plugin to control SDPA behavior:
+    // - CPU plugin disables SDPAFusion (doesn't want SDPA nodes)
```
Comment (v-Golubev) [inline src/common/transformations/src/transformations/common_optimizations/common_optimizations.cpp]: Please note: CPU doesn't always need to decompose SDPA layers. For KV cache scenarios, SDPA is used as a basis for SDPAWithKVCache operation fusion (see `SDPASubgraphFusion` pass in cpu transformation pipeline). And this fusion currently happens before `CommonOptimization` is called, so this change will break LLM scenarios

**Comment by CuriousPanCake** [comment]
1) If I read the description of your PR correctly:
> This change moves SDPAFusion (and SDPAScaleFusion) from MOCTransformations to CommonOptimizations so they only run during compile_model(), where each plugin can control whether SDPA nodes are created and kept.

there might be a case when SDPA is not fused, and it's not going to be fused at the conversion stage, which is completely not ok for the SDPAToPA case which requires the fused SDPA. Maybe there're other problematic cases.

2)
> but the decomposed graph has a different FP32 computation order, causing accuracy loss that amplifies through transformer layers.

What do you mean? Is the order of operations between fused and unfused SDPA implementations different? How so? This is a strict formula with matrix multiplication where order is crucial.

File: src/common/transformations/src/transformations/common_optimizations/common_optimizations.cpp (hunk 1 above)
Comment (evkotov) [reply src/common/transformations/src/transformations/common_optimizations/common_optimizations.cpp]: Thanks for the review, this is a valid concern. I've revised the approach — instead of moving SDPAFusion out of MOCTransformations, the fix is now in ScaledDotProductAttentionDecomposition.

The problem: when SDPAFusion absorbs the K-side scale into the SDPA node, the subsequent decomposition on CPU produces a different FP32 computation order than the original ONNX graph. Specifically, the original computes (Q * s) @ (K^T * s) (both inputs pre-scaled), but decomposition produces (Q * s * s) @ K^T or (Q * s @ K^T) * s — mathematically equivalent but with different rounding errors that accumulate across 14 attention blocks (max_diff grows from ~1e-5 to 0.91).

The fix: in ScaledDotProductAttentionDecomposition, when Q is already pre-scaled (detected as Multiply(input, scalar_constant)), apply the SDPA scale to K^T instead of Q. This restores the original computation order: (Q * s) @ (K^T * s).

SDPAFusion stays in MOCTransformations, so SDPASubgraphFusion, KV-cache scenarios, and PagedAttention are not affected.

**Comment by evkotov** [comment]
> 1. If I read the description of your PR correctly:
> 
> > This change moves SDPAFusion (and SDPAScaleFusion) from MOCTransformations to CommonOptimizations so they only run during compile_model(), where each plugin can control whether SDPA nodes are created and kept.
> 
> there might be a case when SDPA is not fused, and it's not going to be fused at the conversion stage, which is completely not ok for the SDPAToPA case which requires the fused SDPA. Maybe there're other problematic cases.
> 
> 
> > but the decomposed graph has a different FP32 computation order, causing accuracy loss that amplifies through transformer layers.
> 
> What do you mean? Is the order of operations between fused and unfused SDPA implementations different? How so? This is a strict formula with matrix multiplication where order is crucial.

I updated approach with new description

File: src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp [hunk 2]
```diff
@@ -73,7 +73,7 @@ TEST_F(TransformationTestsF, ScaledDotProductAttentionDecompositionStaticBasic)
 
     {
         const auto scaled_dot_product_attention =
-            scaled_dot_product_attention_decomposition(query, key, value, attention_mask, scale, casual);
+            scaled_dot_product_attention_decomposition(query, key, value, attention_mask, scale, casual, true);
```
Comment (CuriousPanCake) [inline src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp]: is the `scale_after_matmul` parameter always true now?

Comment (mryzhov) [reply src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp]: we should test different combinations (matmul->output, inputA, inputB, matmul->output and inputA, etc)

Comment (evkotov) [reply src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp]: fixed

File: src/common/transformations/src/transformations/op_conversions/scaled_dot_product_attention_decomposition.cpp [hunk 3]
```diff
@@ -153,14 +124,11 @@ std::shared_ptr<ov::Node> ov::pass::ScaledDotProductAttentionDecomposition::deco
...
-    if (can_move_scale_after_matmul(query, k_transposed, scale)) {
-        auto atten = register_new_node<v0::MatMul>(query, k_transposed)->output(0);
-        scaled_atten = register_new_node<v1::Multiply>(atten, scale)->output(0);
-    } else {
-        auto q_scaled = register_new_node<v1::Multiply>(query, scale);
-        scaled_atten = register_new_node<v0::MatMul>(q_scaled, k_transposed)->output(0);
-    }
+    // Apply scale after MatMul(Q, K^T) per SDPA specification:
+    //   attn_weight = Q @ K^T * scale
+    // Scale is scalar or single-element tensor, so Multiply broadcasts safely over [S_q, S_kv].
```
Comment (v-Golubev) [inline src/common/transformations/src/transformations/op_conversions/scaled_dot_product_attention_decomposition.cpp:129]: It looks like `ScaledDotProductAttention::validate_and_infer_types` lacks the check on scalar scale, which makes it possible to create an op which is not aligned with the spec. Could you please add the needed check there?

File: src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp [hunk 4]
```diff
@@ -220,12 +219,12 @@ TEST_F(TransformationTestsF, ScaledDotProductAttentionDecomposition_ScalarScale_
...
 
     {
-        auto ref = scaled_dot_product_attention_decomposition(query, key, value, attention_mask, scale, casual, true);
+        auto ref = scaled_dot_product_attention_decomposition(query, key, value, attention_mask, scale, casual);
         model_ref = std::make_shared<ov::Model>(OutputVector{ref}, ParameterVector{query, key, value, attention_mask});
     }
 }
 
-TEST_F(TransformationTestsF, ScaledDotProductAttentionDecomposition_DynamicScale_MultiplyBeforeMatMul) {
+TEST_F(TransformationTestsF, ScaledDotProductAttentionDecomposition_DynamicScale_MultiplyAfterMatMul) {
```
Comment (v-Golubev) [inline src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp]: Minor: `MultiplyAfterMatMul` postfix in all test cases can be removed now I suppose, since the mul is always after

File: src/common/transformations/src/transformations/op_conversions/scaled_dot_product_attention_decomposition.cpp (hunk 3 above)
Comment (mryzhov) [reply src/common/transformations/src/transformations/op_conversions/scaled_dot_product_attention_decomposition.cpp:129]: we have it already - https://github.com/mryzhov/openvino/blob/sdpa_validation_fix/src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp#L96, the issue was only in the decomposition, where the scale factor was restored in the wrong place.

File: src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp (hunk 4 above)
Comment (mryzhov) [inline src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp]:
```suggestion
TEST_F(TransformationTestsF, ScaledDotProductAttentionDecomposition_DynamicScale) {
```

File: src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp [hunk 5]
```diff
@@ -361,6 +353,49 @@ const std::shared_ptr<ov::Node> scaled_dot_product_attention_decomposition(std::
     return result;
 }
 
+TEST_F(TransformationTestsF, ScaledDotProductAttentionDecomposition_PreScaledQuery_MultiplyAfterMatMul) {
```
Comment (mryzhov) [inline src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp]:
```suggestion
TEST_F(TransformationTestsF, ScaledDotProductAttentionDecomposition_PreScaledQuery) {
```

File: src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp [hunk 6]
```diff
@@ -361,6 +353,49 @@ const std::shared_ptr<ov::Node> scaled_dot_product_attention_decomposition(std::
...
+    auto q_scale_const = v0::Constant::create(element::f32, Shape{}, {0.353553f});
+    auto query_prescaled = std::make_shared<v1::Multiply>(raw_query, q_scale_const);
+
+    // SDPA scale = the K-side scale that was absorbed by SDPAFusion
+    auto sdpa_scale = v0::Constant::create(element::f32, Shape{}, {0.353553f});
+
+    const bool casual = false;
+
+    {
+        auto sdp = std::make_shared<v13::ScaledDotProductAttention>(query_prescaled,
```
Comment (CuriousPanCake) [inline src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp]:
```suggestion
        auto sdpa = std::make_shared<v13::ScaledDotProductAttention>(query_prescaled,
```

File: src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp [hunk 7]
```diff
@@ -361,6 +353,49 @@ const std::shared_ptr<ov::Node> scaled_dot_product_attention_decomposition(std::
...
+                                                                    key,
+                                                                    value,
+                                                                    attention_mask,
+                                                                    sdpa_scale,
+                                                                    casual);
+        model = std::make_shared<ov::Model>(OutputVector{sdp}, ParameterVector{raw_query, key, value, attention_mask});
```
Comment (CuriousPanCake) [inline src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp]:
```suggestion
        model = std::make_shared<ov::Model>(OutputVector{sdpa}, ParameterVector{raw_query, key, value, attention_mask});
```

File: src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp [hunk 8]
```diff
@@ -361,6 +353,49 @@ const std::shared_ptr<ov::Node> scaled_dot_product_attention_decomposition(std::
...
+
+    const bool casual = false;
+
+    {
+        auto sdpa = std::make_shared<v13::ScaledDotProductAttention>(query_prescaled,
+                                                                    key,
+                                                                    value,
+                                                                    attention_mask,
+                                                                    sdpa_scale,
+                                                                    casual);
```
Comment (CuriousPanCake) [inline src/common/transformations/tests/op_conversions/scaled_dot_product_decomposition_test.cpp]:
```suggestion
        auto sdpa = std::make_shared<v13::ScaledDotProductAttention>(query_prescaled,
                                                                     key,
                                                                     value,
                                                                     attention_mask,
                                                                     sdpa_scale,
                                                                     casual);
```
